docs/guide/ethereum-provider.md

```diff
@@ -112,83 +173,359 @@ params: [
+In general, you can assume that the MetaMask provider is connected and has emitted this event by the time you are able to reference it.
+
+### disconnect
+
+```typescript
+ethereum.on('disconnect', handler: (error: ProviderRpcError) => void);
+```
+
+The MetaMask provider emits this event if it becomes unable to submit RPC requests to any chain.
+In general, this will only happen due to network connectivity issues or some unforeseen error.
+
+Once `disconnect` has been emitted, MetaMask will not accept any new requests until until `connect` is emitted, which may require reloading the page.
+
+### accountsChanged
+
+```typescript
+ethereum.on('accountsChanged', handler: (accounts: Array<string>) => void);
+```
+
+The MetaMask provider emits this event whenever the return value of the `eth_accounts` RPC method changes.
+`eth_accounts` returns an array that is either empty or contains a single account address that the user decided to expose to the requesting domain (in the case of a website, identified by the _hostname_ of its URL).
+
+This means that `accountsChanged` will be emitted whenever the user's exposed account address changes.
+
+::: tip
+We plan to allow the `eth_accounts` array to be able to contain multiple addresses in the near future.
+:::
+
+### chainChanged
+
+::: warning
+**NOTE:** See the [Chain IDs section](#chain-ids) for important information about the MetaMask provider's chain IDs.
+:::
+
+```typescript
+ethereum.on('chainChanged', handler: (chainId: string) => void);
+```
+
+The MetaMask provider emits this event when the currently connected chain changes.
+
+All RPC requests are be submitted to the currently connected chain.
+Therefore, it's critical to keep track of the current chain ID by listening for this event.
+
+We _strongly_ recommend reloading the page on chain changes, unless absolutely necessary not to.
 
 ```javascript
-ethereum.sendAsync(
-  {
-    method: 'eth_requestAccounts',
-  },
-  (error, response) => {
-    if (error) {
-      // Handle error. Likely the user rejected the login
-      console.error(error);
-    } else {
-      const accounts = response.result;
-      // You now have an array of accounts!
-      // Currently only ever one, e.g.:
-      // ['0xFDEa65C8e26263F6d9A1B5de9555D2931A33b825']
-    }
-  }
-);
+ethereum.on('chainChanged', (_chainId) => window.location.reload());
 ```
 
-### ethereum.enable() (DEPRECATED)
+### message
+
+```typescript
+interface ProviderMessage {
+  type: string;
+  data: unknown;
+}
+
+ethereum.on('message', handler: (message: ProviderMessage) => void);
+```
+
+The MetaMask provider emits this event when it receives some message that the consumer should be notified of.
+The kind of message is identified by the `type` string.
+
+One prominent example of such messages include updates from RPC subscriptions, such as `eth_subscribe`. For such messages, the message `type` will be `eth_subscription`.
```

Comment:
The grammar for this seemed a bit off? It seemed like it was suggesting that `eth_subscription` was the `type` of _any_ "RPC subscriptions" rather than `eth_subscribe` specifically, which is obviously what was meant.
```suggestion
One prominent example is `eth_subscribe` updates. Each update for an `eth_subscribe` RPC subscription will be a message with a `type` of `eth_subscription`.
```

Reply:
ffa34ef